pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,242 @@
+// policy database (rpdb) for ENI IPAM mode.
+//
+// We rely on the ability to fetch the CiliumNode resource because we need to
+// fetch the number associated with the ENI device. The CiliumNode resource
+// contains this information in the Status field.
+//
+// This function should be invoked before exposing the Cilium API because we
+// want to ensure that no workloads are scheduled while this modification is
+// taking place. This migration is related to a bug
+// (https://github.com/cilium/cilium/issues/14336) where an ENI has an ifindex
+// that equals the main routing table number (253-255), causing the rules and
+// routes to be created using the wrong table ID, which could end up
+// blackholing most traffic on the node.
+func MigrateENIRoutingTable(getter ciliumNodeGetter) error {
+	if option.Config.IPAM != option.IPAMENI {
+		return nil
+	}
+
+	rules, err := netlink.RuleList(netlink.FAMILY_V4)
+	if err != nil {
+		return err
+	}
+
+	for _, r := range rules {
+		switch r.Priority {
+		case linux_defaults.RulePriorityEgress:
+			routes, err := netlink.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+				Table: r.Table,
+			}, netlink.RT_FILTER_TABLE)
+			if err != nil {
+				return err
+			}
+
+			// If there are no routes under the same table as the rule, then
+			// skip.
+			if len(routes) == 0 {
+				continue
+			}
+
+			oldTable := r.Table
+
+			// It is sufficient to grab the first route that matches because we
+			// are assuming all routes created under a rule will have the same
+			// ifindex (LinkIndex).
+			newTable, err := RetrieveTableIDFromIfIndex(routes[0].LinkIndex, getter)
+			if err != nil {
+				return err
+			}
+
+			if err := migrateOnUpgrade(oldTable, newTable); err != nil {
+				return err
+			}
+
+		case linux_defaults.RulePriorityEgressv2:
+			// Downgrade case. Unsupported for now.
+		}
+	}
+
+	return nil
+}
+
+func RetrieveTableIDFromIfIndex(ifindex int, getter ciliumNodeGetter) (int, error) {
+	link, err := netlink.LinkByIndex(ifindex)
+	if err != nil {
+		return -1, err
+	}
+
+	cn, err := getter.Get("", v1.GetOptions{})
+	if err != nil {
+		return -1, err
+	}
+
+	mac := link.Attrs().HardwareAddr
+
+	var eni v2.ENI
+	for _, e := range cn.Status.ENI.ENIs {
+		if e.MAC == mac.String() {
+			eni = e
+			break
+		}
+	}
+
+	return linux_defaults.RouteTableInterfacesOffset + eni.Number, nil
+}
+
+// migrateOnUpgrade migrates the ENI routing table for the upgrade path. It
+// accepts the old table ID for the existing rules to migrate. `new` is the new
+// table ID that the rules should be referring to.
+func migrateOnUpgrade(old, new int) error {
+	if old == new {
+		// Nothing to do.
+		return nil
+	}
+
+	if err := replaceDefaultRoutes(old, new); err != nil {
+		return err
+	}
+	if err := createNewRules(
+		linux_defaults.RulePriorityEgress,
+		old,
+		linux_defaults.RulePriorityEgressv2,
+		new,
+	); err != nil {
+		return err
+	}
+	if err := deleteOldRules(linux_defaults.RulePriorityEgress); err != nil {
```

Comment:
When updating the routes or creating new rules above, we only replace the ones with the `old` table number. But now for deleting the old rules (+routes below), this is removing all rules/routes at the priority regardless of table number. This seems a bit inconsistent and could theoretically undo some of the careful ordering guarantees we're trying to make here for hitless upgrade.

Reply:
Good call, will update this.